### DISCUSSION TITLE
what is the proper way to have both mmc source and a2dp sink active together (but used exclusively)

### DISCUSSION THREAD

**Discussion by 0x0fe** [pschatzmann/arduino-audio-tools#856] [Q&A] [answered]
During most part of the device operation i use sdmmc as source (with the player api), however, whenever A2DP sink is connected i want it to be the active "player", what is the proper way to do this? Is there any project in the examples? (i did not find).
Thanks.

**Comment by pschatzmann** [comment, marked as answer]
What have you tried ? 
I don't quite understand your question: All the necessary functionality is available.
Did you have a look at the available methods of BluetoothA2DPSink ? There are callback methods to react on changes of the connection state or audio state...

I am just not sure if there is enough RAM for decoding mp3, after a2dp has been started, so you might need to have a processor with spiram or play wav files

**Comment by 0x0fe** [comment]
well, finally i just put both objects together, sharing the i2c object and it did not cause any problem, and both the sd player and a2dp sink can be used without any problem, which is nice and convenient.

**Comment by 0x0fe** [comment]
n.b yes i use an esp32d0wdr2 which has 2MB of SPIRAM, although i did not check how much of it has been allocated by audio-tools, what i can say is that the IDF itself does not take adavantage of the SPIRAM on the wifi and bluetooth stack much, despite i changed the menuconfig options and recompiled all core libraries. When we set all setting to use SPIRAM in priority it does only allocate about 50KB on it for bluetooth and wifi stacks.

**Comment by pschatzmann** [comment]
I was mainly thinking about the RAM use of the codecs. Did you already try [this](https://www.pschatzmann.ch/home/2022/05/30/esp32-and-psram/)?

**Comment by 0x0fe** [comment]
oh, no, thank's,  i was not aware of this, i will try it for sure
